-Clink-dead-code and -Cinstrument-coverage causes symbol already defined compiler error in nightly

<!--
Thank you for filing a regression report! 🐛 A regression is something that changed between versions of Rust but was not supposed to.

Please provide a short summary of the regression, along with any information you feel is relevant to replicate it.
-->

### Code

With just the following cargo.toml

```toml
[package]
name = "debug-tarp-issue"
version = "0.1.0"
edition = "2018"

[dependencies]
reqwest = { version = "0.12", features = ["blocking"] }
```

and a `main.rs` with the default contents. Running 

```
RUSTFLAGS="-Clink-dead-code -Cinstrument-coverage" cargo test
```

Results in:

```
error: symbol `_RNCNvNtNtNtNtCs6KohR1qTMkp_10hyper_util6client6legacy7connect4http7connect0Bb_` is already defined
   --> /Users/runner/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/hyper-util-0.1.10/src/client/legacy/connect/http.rs:795:8
    |
795 |     Ok(async move {
    |        ^^^^^^^^^^

error: could not compile `hyper-util` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
Error: Process completed with exit code 101.
```

I've done a bit of work to narrow this down. So either of the flags on their own is fine, it's just the combination that causes the issue.,

I also have a MRE reproduction repo here: https://github.com/xd009642/debug-tarp-issue 

### Version with regression

So because this happened in a CI job for tarpaulin and we have some flaky tests and I don't use a mac myself I don't have an exact version. But this first occurred in our CI January 13th 2025.  This is evident because the mac nightly builds have failed from that date until now. Mac stable and beta still work fine.

--- a/compiler/rustc_monomorphize/src/collector.rs
+++ b/compiler/rustc_monomorphize/src/collector.rs
@@ -1509,6 +1509,13 @@ impl<'v> RootCollector<'_, 'v> {
                         }
                         _ => unreachable!(),
                     };
+                    let Ok(instance) = self.tcx.try_normalize_erasing_regions(
+                        ty::TypingEnv::fully_monomorphized(),
+                        instance,
+                    ) else {
+                        // Don't ICE on an impossible-to-normalize closure.
+                        return;
+                    };
                     let mono_item = create_fn_mono_item(self.tcx, instance, DUMMY_SP);
                     if mono_item.node.is_instantiable(self.tcx) {
                         self.output.push(mono_item);

--- a/tests/ui/closures/eager-mono-with-normalizable-upvars.rs
+++ b/tests/ui/closures/eager-mono-with-normalizable-upvars.rs
@@ -1,0 +1,19 @@
+//@ compile-flags: -Clink-dead-code -Csymbol-mangling-version=v0
+//@ build-pass
+
+// Ensure that when eagerly collecting `test::{closure#0}`, we don't try
+// collecting an unnormalized version of the closure (specifically its
+// upvars), since the closure captures the RPIT `opaque::{opaque#0}`.
+
+fn opaque() -> impl Sized {}
+
+fn test() -> impl FnOnce() {
+    let opaque = opaque();
+    move || {
+        let opaque = opaque;
+    }
+}
+
+fn main() {
+    test()();
+}

